apptail/storage/tail_tracker.go

```diff
@@ -0,0 +1,258 @@
+package storage
+
+import (
+	"fmt"
+	"github.com/ActiveState/log"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type Tracker interface {
+	Update(instKey string, childKey string, childVal int64)
+	LoadTailers()
+	Remove(key string)
+	RegisterInstance(instKey string)
+	InitializeChildNode(instKey string, childkey string, offSet int64)
+	Commit() error
+	StartSubmissionTimer(persistInterval time.Duration)
+	IsInstanceRegistered(instKey string) bool
+	IsChildNodeInitialized(instKey string, childkey string) bool
+	GetFileCachedOffset(instkey string, fname string) int64
+	getBuffer() ([]byte, error)
+	CleanUp(clenups map[string]bool)
+}
+
+type BoxedInt64 struct{ V int64 }
+
+type TailNode map[string]*BoxedInt64
+
+type Tailer struct {
+	Instances map[string]TailNode
+}
+
+type tracker struct {
+	storage       Storage
+	Cached        *Tailer // do not expose this, it should ONLY be updated via Tracker methods
+	mux           *sync.Mutex
+	timerStopChan chan struct{} // used to send quit signal to time
+}
+
+var (
+	MinIOTicker = 5 * time.Second
+)
+
+func NewTracker(s Storage) Tracker {
+
+	return &tracker{
+		storage: s,
+		mux:     &sync.Mutex{},
+		Cached: &Tailer{
+			Instances: make(map[string]TailNode),
+		},
+		timerStopChan: make(chan struct{}),
+	}
+}
+
+func (t *tracker) StartSubmissionTimer(persistInterval time.Duration) {
+	if persistInterval.Seconds() <= MinIOTicker.Seconds() {
+		seconds := persistInterval / (1000 * time.Millisecond)
+		log.Warnf("IMPORTANT: Setting tail persist interval to %ds will increase your IO Rate", seconds)
+
+	}
+	ticker := time.NewTicker(persistInterval)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				err := t.Commit()
+				if err != nil {
+					log.Fatal(err)
+
+				}
+			case <-t.timerStopChan:
+				ticker.Stop()
+				return
+			}
+
+		}
+	}()
+}
+
+func (t *tracker) RegisterInstance(instKey string) {
+	t.mux.Lock()
+	if _, instance_exist := t.Cached.Instances[instKey]; !instance_exist {
+		t.Cached.Instances[instKey] = TailNode{}
+		t.formatMap("Current Status")
+	}
+	t.mux.Unlock()
+}
+
+// this is mainly used for testing since we are not exposing Cached via interface
+func (t *tracker) IsInstanceRegistered(instKey string) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
```

Comment:
Please use ```defer``` everywhere. Deferred functions are run not only on return but also during panics. It's possible if it's not already so that someone in future might handle a panic on a go routine and now they're in a deadlock situation because defer wasn't used on the mutex unlock. Defer is the more idiomatic approach with Go.